instance/models/openedx_instance.py

```diff
@@ -0,0 +1,178 @@
+    @property
+    def appserver_set(self):
+        """
+        Get the set of OpenEdxAppServers owned by this instance.
+        """
+        return self.ref.openedxappserver_set
+
+    def set_appserver_active(self, appserver_id):
+        """
+        Mark the AppServer with the given ID as the active one.
+        """
+        app_server = self.appserver_set.get(pk=appserver_id)  # Make sure the AppServer is owned by this instance
+        self.active_appserver = app_server
+        self.save()
+        self.logger.info('Updating DNS: LMS at %s...', self.domain)
+        gandi.set_dns_record(type='A', name=self.sub_domain, value=self.active_appserver.server.public_ip)
+        self.logger.info('Updating DNS: Studio at %s...', self.studio_domain)
+        gandi.set_dns_record(type='CNAME', name=self.studio_sub_domain, value=self.sub_domain)
+
+    @log_exception
+    def spawn_appserver(self):
+        """
+        Provision a new AppServer. If it completes successfully, mark it as active.
+        """
+        # Provision external databases:
+        if not self.use_ephemeral_databases:
+            # TODO: Use db row-level locking to ensure we don't get any race conditions when creating these DBs.
+            # Use select_for_update(nowait=True) to lock this object's row, then do these steps, then refresh_from_db
+            self.logger.info('Provisioning MySQL database...')
+            self.provision_mysql()
+            self.logger.info('Provisioning MongoDB databases...')
+            self.provision_mongo()
+            self.logger.info('Provisioning Swift container...')
+            self.provision_swift()
+
+        config_fields = [field.name for field in OpenEdXAppConfiguration._meta.fields if field.name not in ('id', )]
+        config_copy = {field_name: getattr(self, field_name) for field_name in config_fields}
+
+        app_server = self.appserver_set.create(
+            # Name for the app server: this should usually generate a unique name (and won't cause any issues if not):
+            name="AppServer {}".format(self.appserver_set.count() + 1),
+            # Copy the current value of each setting into the AppServer, preserving it permanently:
+            configuration_database_settings=self.get_database_settings(),
+            **config_copy
+        )
+        if app_server.provision():
+            # If the AppServer provisioned successfully, make it the active one:
+            # Note: if I call spawn_appserver() twice, and the second one provisions sooner, the first one may then
+            # finish and replace the second as the active server. We are not really worried about that for now.
+            self.set_appserver_active(app_server.pk)
+
+
+class SandboxInstance(BaseOpenEdXInstance, GitHubInstanceMixin):
+    """
+    SandboxInstance: represents an Open edX instance created for a pull request.
+    """
+
+    class Meta:
+        verbose_name = 'Open edX Sandbox Instance'
+
+    def update_from_pr(self, pr):
```

Comment:
Most of the logic is in the `AppServer`/`OpenEdXAppServer` classes as well as `BaseOpenEdXInstance`. So 99% of  the logic is the same. The only logic that I separated out here is the code and fields related to pull requests, because I didn't think it makes sense to have the `update_from_pr` code on a production instance, nor bother storing fields like `default_fork`, `github_pr_url`, etc. The only fields a production instance needs are `edx_platform_repository_url` and `edx_platform_commit` (which can just be a branch name like "named-release/dogwood.rc" or "st-gallen"). IMHO all the other logic related to pull requests and splitting those two fields up into detailed sub-fields like "organization" and "commit_id" makes sense for sandboxes but only makes the production code more complex than necessary.

However, if you want me to keep those on the same class, I will. I don't have a strong opinion on this.